### PR TITLE
Pytorch2onnx which can be used for DIM and GCA model

### PR DESCRIPTION
**Pull request by drcut**
This PR support exporting DIM and GCA into ONNX format and comparing the execution results between Pytorch and ONNX.
However, as there is some bug in ONNX runtime, the results between Pytorch and ONNX for DIM may be different

### CODE REVIEW

File: mmedit/models/common/gca_module.py [hunk 1]
```diff
@@ -341,8 +341,10 @@ def pad(self, x, kernel_size, stride):
 
     def get_self_correlation_mask(self, img_feat):
         _, _, h, w = img_feat.shape
+        # As ONNX does not support dynamic num_classes, we have to convert it
+        # into an integrate
```
Comment (hejm37) [inline mmedit/models/common/gca_module.py:345]: `integrate` -> `integer`

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py [hunk 2]
```diff
@@ -77,5 +112,6 @@ def forward(self, inputs):
         out = self.max_unpool2d(out, max_idx_1)
 
         out = self.relu(self.deconv1_1(out))
+
```
Comment (hejm37) [inline mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:115]: Is this line necessary?

File: mmedit/models/mattors/base_mattor.py [hunk 3]
```diff
@@ -129,7 +129,6 @@ def restore_shape(self, pred_alpha, meta):
         """
         ori_trimap = meta[0]['ori_trimap'].squeeze()
         ori_h, ori_w = meta[0]['merged_ori_shape'][:2]
-
```
Comment (hejm37) [inline mmedit/models/mattors/base_mattor.py:132]: Is this change necessary?

File: mmedit/models/mattors/dim.py [hunk 4]
```diff
@@ -67,7 +67,6 @@ def _forward(self, x, refine):
             pred_refine = self.refiner(refine_input, raw_alpha)
         else:
             pred_refine = None
-
```
Comment (hejm37) [inline mmedit/models/mattors/dim.py:70]: Same as above~

File: mmedit/models/mattors/gca.py [hunk 5]
```diff
@@ -91,8 +91,9 @@ def forward_test(self,
             dict: Contains the predicted alpha and evaluation result.
         """
         pred_alpha = self._forward(torch.cat((merged, trimap), 1))
-
-        pred_alpha = pred_alpha.cpu().numpy().squeeze()
+        if torch.onnx.is_in_onnx_export():
```
Comment (hejm37) [inline mmedit/models/mattors/gca.py:94]: Since there is already a `forward_dummy()`, maybe you can make use of that?

Comment (drcut) [reply mmedit/models/mattors/gca.py]: OK, I have submitted a new commit which use forward_dummy for onnx exporting

File: mmedit/models/mattors/dim.py (hunk 4 above)
Comment (drcut) [reply mmedit/models/mattors/dim.py:70]: > Same as above~

Sorry, it's my fault, I have retrieved this change.

File: mmedit/models/mattors/base_mattor.py (hunk 3 above)
Comment (drcut) [reply mmedit/models/mattors/base_mattor.py]: > Is this change necessary?

Sorry, it's my fault, I have retrieved this change.

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py (hunk 2 above)
Comment (drcut) [reply mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py]: > Is this line necessary?

Sorry, it's my fault, I have retrieved this change.

Comment (drcut) [reply mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py]: > Is this line necessary?

Sorry, it's my fault, I have retrieved this change.

File: mmedit/models/common/gca_module.py (hunk 1 above)
Comment (drcut) [reply mmedit/models/common/gca_module.py]: > `integrate` -> `integer`

Done

Comment (drcut) [reply mmedit/models/common/gca_module.py]: > `integrate` -> `integer`

Done

Comment (drcut) [reply mmedit/models/common/gca_module.py]: > `integrate` -> `integer`

Done

File: tools/pytorch2onnx.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            onnx_result = onnx_result[0]
+        # check the numerical value
+        assert np.allclose(
+            pytorch_result,
+            onnx_result), 'The outputs are different between Pytorch and ONNX'
+        print('The numerical values are same between Pytorch and ONNX')
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Convert MMDet to ONNX')
```
Comment (hejm37) [inline tools/pytorch2onnx.py:65]: Rename `MMDet` to `MMEdit` accordingly?

Comment (drcut) [reply tools/pytorch2onnx.py:65]: > Rename `MMDet` to `MMEditing` accordingly?

Ok, I will fix it.
BTW, as I have just decided to move symblic.py and other file to mmcv, so I will submit a new commit after my corresponding PR merged to MMCV

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py [hunk 7]
```diff
@@ -1,6 +1,46 @@
...
+import torch.nn.functional as F
 from mmcv.cnn.utils.weight_init import xavier_init
 from mmedit.models.registry import COMPONENTS
+from torch.autograd import Function
+from torch.nn.modules.pooling import _MaxUnpoolNd
+from torch.nn.modules.utils import _pair
+
+
+class MaxUnpool2dop(Function):
+
```
Comment (hejm37) [inline mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:13]: Maybe you need to add docstring to explain the added classes and functions.

File: tools/pytorch2onnx.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+from mmedit.models import build_model
+
+
+def pytorch2onnx(model,
+                 input,
+                 opset_version=11,
+                 show=False,
+                 output_file='tmp.onnx',
+                 verify=False):
+    model.cpu().eval()
```
Comment (hejm37) [inline tools/pytorch2onnx.py:20]: Docstring~

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py [hunk 9]
```diff
@@ -1,6 +1,69 @@
...
+    @staticmethod
+    def forward(ctx, input, indices, kernel_size, stride, padding,
+                output_size):
+        return F.max_unpool2d(input, indices, kernel_size, stride, padding,
+                              output_size)
+
+    @staticmethod
+    def symbolic(g, input, indices, kernel_size, stride, padding, output_size):
+        warnings.warn(
```
Comment (hejm37) [inline mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:26]: Docstring should be also added to these functions.

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py [hunk 10]
```diff
@@ -1,6 +1,69 @@
...
+
+class MaxUnpool2dop(Function):
+    """We warp the `torch.nn.functional.max_unpool2d`
+    with an extra `symbolic` method, which is needed while exporting to ONNX.
+    Users should not call this function directly.
+    """
+
+    @staticmethod
+    def forward(ctx, input, indices, kernel_size, stride, padding,
+                output_size):
```
Comment (hejm37) [inline mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:20]: Docstring should be also added to these functions.

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py [hunk 11]
```diff
@@ -1,6 +1,69 @@
...
+    """
+
+    def __init__(self, kernel_size, stride=None, padding=0):
+        super(MaxUnpool2d, self).__init__()
+        self.kernel_size = _pair(kernel_size)
+        self.stride = _pair(stride or kernel_size)
+        self.padding = _pair(padding)
+
+    def forward(self, input, indices, output_size=None):
+        """Forward function of PlainDecoder.
```
Comment (hejm37) [inline mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:54]: Docstring should be also added to these functions.

File: tools/pytorch2onnx.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+
+def pytorch2onnx(model,
+                 input,
+                 opset_version=11,
+                 show=False,
+                 output_file='tmp.onnx',
+                 verify=False):
+    """Export Pytorch model to ONNX model and verify the outputs are same
+    between Pytorch and ONNX
```
Comment (hejm37) [inline tools/pytorch2onnx.py:21]: `ONNX.`

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py (hunk 9 above)
Comment (drcut) [reply mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:26]: > Docstring should be also added to these functions.

Maybe we do not need to add docstring for this function, as this function can not be called by users, but only be invoked by Pytorch engine while exporting to ONNX.

File: tools/pytorch2onnx.py (hunk 12 above)
Comment (drcut) [reply tools/pytorch2onnx.py:21]: > `ONNX.`

Done

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py (hunk 9 above)
Comment (hejm37) [reply mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:39]: All right.

File: mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py [hunk 13]
```diff
@@ -1,6 +1,82 @@
...
+
+
+class MaxUnpool2d(_MaxUnpoolNd):
+    """This module is modified from Pytorch `MaxUnpool2d` module.
+
+    Args:
+      kernel_size (int or tuple): Size of the max pooling window.
+      stride (int or tuple): Stride of the max pooling window.
+          It is set to `kernel_size` by default.
+      padding (int or tuple): Padding that is added to the input.
```
Comment (hejm37) [inline mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:57]: Default value of `padding`~

Comment (hejm37) [reply mmedit/models/backbones/encoder_decoders/decoders/plain_decoder.py:57]: You can also pull the master branch to resolve the conflict~
